src/main/java/hudson/remoting/RemoteClassLoader.java

```diff
@@ -1124,6 +1121,88 @@ private Object readResolve() throws ObjectStreamException {
         }
     }
 
+    /**
+     * An {@link InvocationHandler} for proxying calls to a target {@link IClassLoader} which catches
+     * {@link RemotingSystemException} caused by an {@link InterruptedException} (or {@link InterruptedIOException}).
+     * When this happens, the target method is called again, until it successfully returns (or throws a different kind
+     * of exception), or a maximum number of retries has been reached. The current thread is then interrupted again.
+     * <p>
+     * The purpose of this proxy is to avoid interruption of class loading/initialization, because that's something a
+     * class loader can't recover from. Later attempts at using an affected class will result in a {@link LinkageError},
+     * as reported in JENKINS-36991, JENKINS-51854, JENKINS-61103 and others.
+     *
+     * @see <a href="https://issues.jenkins-ci.org/browse/JENKINS-61103">JENKINS-61103</a>
+     */
+    static class RemoteInterruptedExceptionCatcher implements InvocationHandler {
+
+        private final IClassLoader targetProxy;
+
+        public RemoteInterruptedExceptionCatcher(IClassLoader targetProxy) {
+            this.targetProxy = targetProxy;
+        }
+
+        /**
+         * @inheritDoc
+         */
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+            Object result;
+            boolean interrupted = false;
+            int retryCounter = 0;
+            try {
+                do {
+                    try {
+                        result = method.invoke(targetProxy, args);
+                        break;
+                    } catch (InvocationTargetException x) {
+                        Throwable e = x.getCause(); // get the actual exception thrown by the target method call
+                        if (retryCounter++ >= MAX_RETRIES_ON_REMOTING_INTERRUPTION || !shouldRetry(method, e)) {
```

Comment:
How about moving that first part (before the `||`) into the `shouldRetry()` method?

Reply:
505ad3b4

The expression in `shouldRetry` is becoming a bit long maybe, do you think it's time split into `if / else if` with intermediate `return false`?

Reply:
I would give that a try. It is getting kind of long and cumbersome.